### PR TITLE
Migrate ListingDemoPage to TypeScript.

### PR DESCRIPTION
**Pull request by richardxia**
Refs #1175.

This migrates a couple of testing and debug-only files to TypeScript. It was honestly kind of difficult to figure out the exact types, since they're pretty dynamic, so I'm glad that I did this because now it's easier to understand the types of the test data here.

### CODE REVIEW

File: app/pages/debug/ListingDemoPage.tsx [hunk 1]
```diff
@@ -24,29 +28,41 @@ export const simpleScheduleLookup = {
...
-export const createScheduleFromShorthand = shedule_shorthand => {
+/** Return true if x is an Interval[]; false if it is only an Interval.
+ *
+ * This is a user-defined type guard, since TypeScript is unable to infer the
+ * correct type normally.
+ *
+ * For more information on user-defined type guards, see:
+ * https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates
+ */
+function isIntervalArray(x: Interval | readonly Interval[]): x is readonly Interval[] {
```
Comment (davidagustin) [inline app/pages/debug/ListingDemoPage.tsx:44]: I still don't understand why the return type here is not a Boolean.

**Review by davidagustin — CHANGES_REQUESTED** [review]

File: app/pages/debug/ListingDemoPage.tsx (hunk 1 above)
Comment (richardxia) [reply app/pages/debug/ListingDemoPage.tsx:44]: It's just special syntax in TypeScript for user-defined type guards. I linked to the TypeScript documentation on user-defined type guards in the comment for this function. Is there a particular part of that documentation that you wanted a more detailed explanation for?

Comment (davidagustin) [reply app/pages/debug/ListingDemoPage.tsx:44]: Got it. I saw a part of the documentation but there are different kinds of narrowing.

**Review by davidagustin — APPROVED** [review]

**Comment by davidagustin** [comment]
Learnt more typescript today. LGTM

**Comment by schroerbrian** [comment]
Looks like a robust and helpful change. Thank you!

**Pull request merged**